File: FS/Pre-FS_programs/Day19-16Nov_2021/program2.cpp
/*
There are n boxes in a row. row[i] is the number printed on ith box(0-indexed).
Your task is to find the smallest index  I of the row, 
where row[I] equals to the remainder when you divide I with 10.

If no such index found, print -1.

Constraints:
    1 <= n <= 100
    0 <= row[i] <= 9
Input Format:
-------------
Line-1: An integer n, number of boxes.
Line-2: n space separated integers represent numbers printed on the boxes
in the row.
 
Output Format:
--------------
Print an integer, smallest index possible.
 
 
Sample Input-1:
---------------
5
2 1 2 3 5
 
Sample Output-1:
----------------
1
 
Sample Input-2:
---------------
5
4 3 1 5 2

Sample Output-2:
----------------
-1*/

#include <iostream>
using namespace std;

int findSmallest(int array[], int length){
    // Check-1...
    for(int i=0; i<length; i++)
        if(array[i] == array[i]%10      )
            continue;
        else   
            return -1;
    cout << "Check-1 passed..!!" << endl;
    // Check-2... control comes here only if passed check-1...
    for(int i=0; i<length; i++)   
        if(array[i] == i)
            return i;
    return -1;
}

int main()
{
    int length=0, temp=0;
    cin >> length;
    int array[length]={0};
    for(int i=0; i<length; i++){
        cin >> array[i];
    }

    cout << findSmallest(array, length);
}

File: FS/Pre-FS_programs/Day16-2ndNov_2021/program2.java
/*
Chhota Bheem is fond of Laddus, there are N Laddus in a row, N is an even number. 
Where k-th laddu is of type-laddu[k], 1<=k<=N.
 
Due to health issues, Chhota Bheem was suggested to eat atmost N/2 Laddus of 
different types. He can eat only 1 Laddu of each type.
 
You are given an list of integers,
Your task is to find the maximum number of Laddus, Chhota Bheem can eat.
 
 
Input Format:
-------------
Line -1:An integer N,represents number of Lddus.
Line-2: N space separated integers, Laddu types.
 
Output Format:
--------------
Print an integer result.
 
Sample Input-1:
---------------
6
2 4 1 2 3 4
 
Sample Output-1:
----------------
3
 
Explanation:
------------
There are 6 Laddus, and 4 types of Laddus, 
So Bheem can eat 3 laddus only.
 
 
Sample Input-2:
---------------
8
1 1 1 2 1 2 1 1
 
Sample Output-2:
----------------
2
 
Explanation:
------------
There are 8 Laddus, and 2 types of Laddus, 
So Bheem can eat 2 ladddus only.
*/

import java.util.*;

class numLaddos{

    public static int maxLaddosBheemCanEat(int[]laddus){
        HashSet<Integer> laddu_types = new HashSet<Integer>();
        for(int laddu: laddus)
            laddu_types.add(laddu);

        if(laddu_types.size() > laddus.length/2)
            return laddus.length/2;
        else
            return laddu_types.size();
    }

    public static void main(String [] args){
        Scanner scanner = new Scanner(System.in);
        int laddos_count = scanner.nextInt();
        int laddos[] = new int[laddos_count];
        for(int i=0; i<laddos_count; i++)
            laddos[i] = scanner.nextInt();

        System.out.println(maxLaddosBheemCanEat(laddos));
    }
}

/*
+------------------------------+
|  8 tests run/ 8 tests passed |
+------------------------------+
 *
 * */
